sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenProviderWithSharedAccessKey.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.Iot.Hub.Service.Authentication
+{
+    /// <summary>
+    /// Implementation of a shared access signature provider with token caching and refresh.
+    /// </summary>
+    internal class SasTokenProviderWithSharedAccessKey : ISasTokenProvider
+    {
+        private static readonly TimeSpan s_defaultTimeToLive = TimeSpan.FromMinutes(30);
+        private readonly object _lock = new object();
+
+        private readonly string _hostname;
+        private readonly string _sharedAccessPolicy;
+        private readonly string _sharedAccessKey;
+        private readonly TimeSpan _timeToLive;
+
+        private string _cachedSasToken;
+        private DateTimeOffset _tokenExpiryTime;
+
+        // Private constructor to prevent accidental initialzation.
+        private SasTokenProviderWithSharedAccessKey()
+        {
+        }
+
+        public SasTokenProviderWithSharedAccessKey(string hostname, string sharedAccessPolicy, string sharedAccessKey, TimeSpan? timeToLive = null)
+        {
+            _hostname = hostname;
+            _sharedAccessPolicy = sharedAccessPolicy;
+            _sharedAccessKey = sharedAccessKey;
+            _timeToLive = timeToLive ?? s_defaultTimeToLive;
+
+            _cachedSasToken = null;
+        }
+
+        public string GetSasToken()
```

Comment:
Even though this is public, this class is internal. Should a user for sastoken being used how do you plan to handle that?

Reply:
Can you elaborate?

Reply:
I don't understand the question, could you explain it?
The user isn't expected to call GetSasToken() directly, they would only pass in the connection string.

Reply:
I was super unclear (re-reading it now) Let me try again.

I as a user provided you a key. You as SDK built a sastoken. Now I as a user might want this sastoken you built (as I want to reuse it else where)
If user ask us (which happened in the past) to provide a sastoken getter for the key they provided, how should we handle that.
- Can we just say we wont give ? (to check with board on that)
- or
- If we decide to give, how is your design making that happen?

Reply:
I see, right now we do not have any way to expose sas tokens to users, we would need clarification from azure sdk board on that. 

For TokenCredential based authentication, the sdk doesn't really manage the tokens, so it isn't relevant there. For connectionString scenario, if we don't have an equivalent type from Azure.Core and we have to go with our own implementation, we would have to manage that.

Reply:
Tracked here: https://msazure.visualstudio.com/One/_workitems/edit/7462959